docs/async-await.md

```diff
@@ -0,0 +1,168 @@
+
+  /// The request's body
+  var body: Body?
+    
+  init(url: String) {
+    self.url = url
+    self.method = .GET
+    self.headers = .init()
+    self.body = .none
+  }
+}
+```
+
+A notable change from the current [`HTTPRequest`][HTTPRequest] is that the url is not of type `URL`. This makes the creation of a request non throwing. Existing issues regarding current API: 
+
+- [HTTPClient.Request.url is a let constant][issue-395]
+- [refactor to make request non-throwing](https://github.com/swift-server/async-http-client/pull/56)
+- [improve request validation](https://github.com/swift-server/async-http-client/pull/67)
+
+The url validation will become part of the normal request validation that occurs when the request is scheduled on the `HTTPClient`. If the user supplies a request with an invalid url, the http client, will reject the request.
+
+In normal try/catch flows this should not change the control flow:
+
+```swift
+do {
+  var request = AsyncRequest(url: "invalidurl")
+  try await httpClient.execute(request, deadline: .now() + .seconds(3))
+} catch {
+  print(error)
+}
+```
+
+If the library code throws from the AsyncRequest creation or the request invocation the user will, in normal use cases, handle the error in the same catch block. 
+
+#### Body streaming
+
+The new AsyncRequest has a new body type, that is wrapper around an internal enum. This allows us to evolve this type for use-cases that we are not aware of today. 
+
+```swift
+public struct Body {
+  static func bytes<S: Sequence>(_ sequence: S) -> Body where S.Element == UInt8
+
+  static func stream<S: AsyncSequence>(_ sequence: S) -> Body where S.Element == ByteBuffer
+
+  static func stream<S: AsyncSequence>(_ sequence: S) -> Body where S.Element == UInt8
+}
+```
+
+The main difference to today's `Request.Body` type is the lack of a `StreamWriter` for streaming scenarios. The existing StreamWriter offered the user an API to write into (thus the user was in control of when writing happened). The new `AsyncRequest.Body` uses `AsyncSequence`s to stream requests. By iterating over the provided AsyncSequence, the HTTPClient is  in control when writes happen, and can ask for more data efficiently. 
+
+Using the `AsyncSequence` from the Swift standard library as our upload stream mechanism dramatically reduces the learning curve for new users.
+
+### New `AsyncResponse` type
+
+The `AsyncResponse` looks more similar to the existing `Response` type. The biggest difference is again the `body` property, which is now an `AsyncSequence` of `ByteBuffer`s instead of a single optional `ByteBuffer?`. This will make every response on AsyncHTTPClient streaming by default.
+
+```swift
+public struct AsyncResponse {
+  /// the used http version
+  public var version: HTTPVersion
+  /// the http response status
+  public var status: HTTPResponseStatus
+  /// the response headers
+  public var headers: HTTPHeaders
+  /// the response payload as an AsyncSequence
+  public var body: Body
```

Comment:
It might be worth exploring an akka http entity inspired design here before we commit to this API. 
Specially for the response object it matters because as we create it we _know_ if there will be more bytes or not, so we can immediately construct the right "entity object", e.g. Empty or Strict(<all the bytes in mem>) or a stream.

Representing the body as an abstract `HTTPEntity` which may be "strict" or not allows us then to create store the strict bytes via reference to it directly, we don't have to use the stream then at all other than lazy creating it when someone pulls from it.

--- 

The entire topic of "there's a stream of bytes, but noone subscribed to it" is quite hellish in general.

What is the approach of this implementation about it? If we're doing a proper back-pressured stream it means that not reading it is directly connected to not reading from the wire, which may or may not be what the user intended. In akka we had to invent subscription timeouts to resolve such zombie streams.

In case I'm not clear about the case: response has streamed body; we're a nice streaming API and people expect back-pressure via such stream API. If we _never_ consume this stream, what happens?

(In Akka's case we were very loud that this is user error and one has to do `.discardBytes()` (but it's annoying))

--- 

Note though that I don't yet have good performance intuitions about AsyncSequences... perhaps it is not worth optimizing for knowing that there's a `.single(<bytes>)` at all since setup and logic costs are comparable between this and an iterator... So mostly sharing as an "are we _sure_ and have we considered alternatives?" note.

Reply:
> In case I'm not clear about the case: response has streamed body; we're a nice streaming API and people expect back-pressure via such stream API. If we never consume this stream, what happens?
>
> (In Akka's case we were very loud that this is user error and one has to do .discardBytes() (but it's annoying))

Reply:
We talked offline a little bit -- I think we're in a better place here, because the request's dropping can cancel (and does, but was not documented in this document), which leaves us without the nasty "hanging stream" issue 👍 

The must always consume rule is fine, we'll have to be explicit about it 👍